server/lib/actors/hot.go

```diff
@@ -0,0 +1,741 @@
+)
+
+//
+// FetchResult 
+//
+// FetchResult represents data format which fetcher should return
+type FetchResult struct {
+	id proto.UUID		//target id
+	parent proto.UUID	//target parent id. FetchResult will be group by parent
+	score int32			//target score. higher score regard as "hot".
+	vote uint32 		//number of vote (up + down)
+}
+
+
+//
+// entry and its sorter
+//
+// HotEntry represents single hot object with its score and id
+type HotEntry struct {
+	Id proto.UUID
+	Score int32
+	Vote uint32
+}
+
+func (e *HotEntry) within(start, end time.Time) bool {
+	return yue.UUIDBetween(yuep.UUID(e.Id), start, end)
+}
+
+func (e *HotEntry) flamy() bool {
+	if e.Vote < FLAME_VOTE_THRESHOLD {
+		return false
+	}
+	return math.Abs(float64(e.Score) / float64(e.Vote)) < 2 * FLAME_THRESHOLD
+}
+
+func (e *HotEntry) lessThanByScore(score int32, id proto.UUID) bool {
+	if e.Score < score {
+		return true
+	} else if e.Score > score {
+		return false
+	}
+	return e.Id < id
+}
+
+func (e *HotEntry) lessThanByVote(vote uint32, id proto.UUID) bool {
+	if e.Vote < vote {
+		return true
+	} else if e.Vote > vote {
+		return false
+	}
+	return e.Id < id
+}
+
+// hotlist represents a set of HotEntry in certain category 
+type hotlist struct {
+	entries []HotEntry
+	indexes map[proto.UUID]int
+}
+
+func NewHotList(size int) *hotlist {
+	return &hotlist {
+		entries: make([]HotEntry, 0, size),
+		indexes: make(map[proto.UUID]int),
+	}
+}
+
+func NewHotListByEntries(entries []HotEntry) *hotlist {
+	l := NewHotList(len(entries))
+	l.addEntries(entries)
+	return l
+}
+
+func (l *hotlist) addEntries(entries []HotEntry) {
+	for _, e := range entries {
+		l.addEntry(e)
+	}
+}
+
+func (l *hotlist) addEntry(e HotEntry) {
+	i, ok := l.indexes[e.Id]
+	//log.Printf("addEntries: %v %v %v", e, i, len(l.entries))
+	if ok {
+		l.entries[i].Score += e.Score
+		l.entries[i].Vote += e.Vote
+	} else {
+		l.indexes[e.Id] = len(l.entries)
+		l.entries = append(l.entries, e)
+	}
+}
+
+func (l *hotlist) dup() *hotlist {
+	d := NewHotList(len(l.entries))
+	d.addEntries(l.entries)
+	return d
+}
+
+func (l *hotlist) sort(btyp proto.TopicListRequest_BucketType) {
+	switch btyp {
+	case proto.TopicListRequest_Rising, proto.TopicListRequest_Hot:
+		sort.Sort(&ByScore{ l: l })
+	case proto.TopicListRequest_Flame:
+		sort.Sort(&ByVote{ l: l })
+	default:
+		panic(fmt.Sprintf("invalid bucket type %v", btyp))
+	}
+}
+
+func (l *hotlist) indexByScore(btyp proto.TopicListRequest_BucketType, val int32, id proto.UUID) int {
+	if id <= 0 {
+		return 0
+	}
+	f := func (n int) bool {
+		e := l.entries[n]
+		return e.lessThanByScore(val, id)
+	}
+	if btyp == proto.TopicListRequest_Flame {
+		f = func (n int) bool {
+			e := l.entries[n]
+			return e.lessThanByVote(uint32(val), id)
+		}
+	}
+	return sort.Search(len(l.entries), f)
+}
+
+func (l *hotlist) rangeOf(btyp proto.TopicListRequest_BucketType, startScore int32, startId proto.UUID, count uint32) []HotEntry {
```

Comment:
__[gometalinter]__
* [gotype] undeclared name: proto
* [golint] method parameter startId should be startID